### PR TITLE
Add RevealOnDeath trait

### LINKED ISSUE #7916: Delay growing back the shroud when an actor gets killed
So things like demo truck explosions and sandworm attacks will be visible to the player

### PR DESCRIPTION
**Pull request by reaperrr**
...and use it in D2k to reveal las actor position for 4 seconds.

Note: `RevealOnFire` still needs to stay on sandworm, because swallowing doesn't trigger regular death.

Closes #7916.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/RevealOnDeath.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Reveal this actor to the target's owner when killed.")]
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/RevealOnDeath.cs]: it's owner, I presume?

Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/RevealOnDeath.cs]: Yep, copy-pasta, will fix it.

**Comment by reaperrr** [comment]
Updated.

**Comment by GraionDilach** [comment]
Could this have a damagetype setting?

**Comment by penev92** [comment]
I have made a case against arbitrary default values on traits many times in the past. I may have fallen out of the loop and things may be done this way now (again), but I'd still like to raise my concern about this practice. Just my two cents, seeing as I am sadly no longer an active member of the team.

**Comment by GraionDilach** [comment]
@penev92 Arbitrary default values are a problem if they rely on a value based on other actors - Aircraft's RepairableBuildings, for example - but when they work with every possible mod (due to being a global setup), then I don't see where is the problem.

**Comment by SoScared** [comment]
Very exciting! Would this be able to affect the position of aircraft crashes as well?

**Comment by reaperrr** [comment]
> but not with the crash site position(YAK Husk?).

pinging @pchote: The `FallToEarth` activity uses Dispose instead of Kill. I think it's too risky to count a Dispose as 'Death' in my trait, so my question is, would it cause any trouble to change `FallToEarth` activity to use Kill instead?

**Comment by pchote** [comment]
As long as the husk doesn't have `UpdatesPlayerStatistics` or any other traits that implement `INotifyKilled` it *should* (but no guarantees) be fine.  This would also let us fix the TS jumpjet death animation.

**Comment by reaperrr** [comment]
Updated, added a commit that makes `FallToEarth` kill the actor instead of disposing it, allowing aircraft husks to use `RevealOnDeath`.

**Comment by GraionDilach** [comment]
I still request a DamageType setting on this, but :+1: after that.

**Comment by obrakmann** [comment]
Looks good, but I agree that adding a check for `DamageType`s might be useful.

**Comment by reaperrr** [comment]
Updated.
Travis didn't run, but I guess you can `make check/test` to compensate for that.

**Comment by pchote** [comment]
Why not include this on buildings or carryalls?

**Comment by reaperrr** [comment]
> Why not include this on buildings or carryalls?

Updated.

**Review by GraionDilach — APPROVED** [review]
:+1:

File: mods/d2k/rules/aircraft.yaml [hunk 2]
```diff
@@ -51,6 +51,9 @@ carryall:
 	Buildable:
 		Queue: Aircraft
 		BuildPaletteOrder: 120
+	RevealOnDeath:
```
Comment (pchote) [inline mods/d2k/rules/aircraft.yaml]: Sorry, my mistake.  I forgot that carryalls don't have any vision of their own, so revealing on their death looks weird.  Can you drop this again?

**Review by pchote — APPROVED** [review]
LGTM :+1:.  Just one minor issue then we can merge:

**Comment by reaperrr** [comment]
Updated.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/082a66ddb5345bf2619fee3ce42f91c20afda2c7

**Pull request merged**